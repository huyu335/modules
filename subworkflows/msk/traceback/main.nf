// TODO nf-core: If in doubt look at other nf-core/subworkflows to see how we are doing things! :)
//               https://github.com/nf-core/modules/tree/master/subworkflows
//               You can also ask for help via your pull request or on the #subworkflows channel on the nf-core Slack workspace:
//               https://nf-co.re/join
// TODO nf-core: A subworkflow SHOULD import at least two modules
include { PVMAF_CONCAT as PVMAFCONCAT_INITIAL} from '../../../modules/msk/pvmaf/concat'
include { PVMAF_CONCAT as PVMAFCONCAT_GENOTYPE} from '../../../modules/msk/pvmaf/concat'
include { PVMAF_TAG                      } from '../../../modules/msk/pvmaf/tag'
include { GENOTYPEVARIANTS_ALL as GENOTYPEVARIANTS_ALL_ST} from '../../../modules/msk/genotypevariants/all'
include { GENOTYPEVARIANTS_ALL as GENOTYPEVARIANTS_ALL_LQ} from '../../../modules/msk/genotypevariants/all'

workflow TRACEBACK {

    take:
    // TODO nf-core: edit input (take) channels
    bams // channel: [ val(meta), bam, bam.bai ] or [ val(meta), [], [], _duplex.bam, _duplex.bam.bai, _simplex.bam, _simplex.bam.bai ]
    mafs // channel: [ val(meta), [maf] ]
    reference
    reference_fai

    main:

    ch_versions = Channel.empty()
    

    // [[patient:null, id:C-8LCJE9-M001-d], [], [], /Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-duplex.bam, /Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-duplex.bam.bai, /Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-simplex.bam, /Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-simplex.bam.bai]
    // [[patient:null, id:s_C_FWD2FT_X009_d], /Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam, /Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam.bai, [], [], [], []]
    // [[patient:null], [/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d_data_mutations.txt, /Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d_data_mutations_copy.txt, /Users/ebuehler/Downloads/traceback_test/test_data/P-0038647-T04-IM6_data_mutations.txt, /Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d.DONOR22-TP.combined-variants.vep_keptrmv_taggedHotspots_fillout_filtered.maf]]
    // [ [sample: 'sample', patient: 'patient'],maf,bam, bam_xs ]
    // .collect(sample_files)
    // tuple("patient":row.patient_id, "id":row.sample_id, file(row.standard_bam), file(row.standard_bam + '.bai'),[],[],[],[])
    // .map {row -> tuple("patient":row.patient_id, file(row.maf))}
    // .groupTuple(by:0) // group by patient if provided
    // .set{maf_list}
    // reference
    // // Concat Input Mafs, per patient if provided
    PVMAFCONCAT_INITIAL(mafs)
    ch_versions = ch_versions.mix(PVMAFCONCAT_INITIAL.out.versions.first())

    // // genotype --all, make container and module
    bams
    .map { tuple( 'patient': it[0]['patient'], *it ) }
    .combine( PVMAFCONCAT_INITIAL.out.maf, by: 0 )
    .map { it[1..-1] }
    .set{s_bam_list_maf}

    GENOTYPEVARIANTS_ALL_ST(s_bam_list_maf, reference.collect(), reference_fai.collect())
    ch_versions = ch_versions.mix(GENOTYPEVARIANTS_ALL_ST.out.versions.first())
    // Grab Relevant Maf Output from GENOTYPEVARIANTS_ALL_(ST|LQ)
    // Grab Relevant Maf Output from GENOTYPEVARIANTS_ALL_(ST|LQ)
    GENOTYPEVARIANTS_ALL_ST.out.maf
    .transpose()
    .branch {
        genotyped: it[1] =~ /.*ORG-STD_genotyped.maf/
        }
    .set{split_genotype_st}

    GENOTYPEVARIANTS_ALL_ST.out.maf
    .transpose()
    .branch {
        genotyped: it[1] =~ /.*ORG-SIMPLEX-DUPLEX_genotyped.maf/
        }
    .set{split_genotype_lq}
    split_genotype_lq.view()
    // Combine GENOTYPEVARIANTS_ALL_(ST|LQ) Mafs and Group by Patient
    split_genotype_st
    .concat(split_genotype_lq.genotyped)
    .map{meta, files -> tuple('patient': meta['patient'], files )}
    .groupTuple()
    .set{all_genotype}
    // // Concat Input Mafs, per patient if provided
    // // These combine the two stats from standard and access.
    // // pv maf traceback combine
    PVMAFCONCAT_GENOTYPE(all_genotype)
    ch_versions = ch_versions.mix(PVMAFCONCAT_GENOTYPE.out.versions.first())
    // Tag with Traceback Columns
    PVMAF_TAG(PVMAFCONCAT_GENOTYPE.out.maf, 'traceback')
    ch_versions = ch_versions.mix(PVMAF_TAG.out.versions.first())
    PVMAF_TAG.out.maf.view()

    emit:
    maf = PVMAF_TAG.out.maf
    versions = ch_versions                     // channel: [ versions.yml ]
}

