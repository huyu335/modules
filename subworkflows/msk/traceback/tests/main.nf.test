// nf-core subworkflows test traceback
nextflow_workflow {

    name "Test Subworkflow TRACEBACK"
    script "../main.nf"
    workflow "TRACEBACK"
    config "./nextflow.config"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows_msk"
    tag "subworkflows/traceback"
    tag "genotypevariants/all"
    tag "pvmaf/concat"
    tag "pvmaf/tag"
    

        test("sarscov2 - bam - single_end") {

        when {
            workflow {
                """
                // TODO nf-core: define inputs of the workflow here. Example:

                input[0] = Channel.of(
                    [[id:'s_C_FWD2FT_X009_d', patient: 'test1'],
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam'), 
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam.bai'), 
                        [], [], [], []
                    ],
                    [[id:'s_C_FWD2FT_X008_d', patient: 'test1'],
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam'), 
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d.rg.md.abra.printreads.bam.bai'), 
                        [], [], [], []
                    ],
                    [[id:'P-0038647-T04-IM6', patient: 'test2'],
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/NK237840-T.bam'), 
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/NK237840-T.bam.bai'), 
                        [], [], [], []
                    ],
                    [[id:'C-8LCJE9-M001-d', patient:'test3'], [], [],
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-duplex.bam'),
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-duplex.bam.bai'), 
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-simplex.bam'),
                        file('/Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d_cl_aln_srt_MD_IR_FX_BR__aln_srt_IR_FX-simplex.bam.bai')
                    ])

                input[1] = Channel.of(
                    [[patient: 'test1'], 
                    [file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d_data_mutations.txt'),
                     file('/Users/ebuehler/Downloads/traceback_test/test_data/s_C_FWD2FT_X009_d_data_mutations_copy.txt')
                    ]
                    ],
                    [[patient: 'test2'], 
                    file('/Users/ebuehler/Downloads/traceback_test/test_data/C-8LCJE9-M001-d.DONOR22-TP.combined-variants.vep_keptrmv_taggedHotspots_fillout_filtered.maf')
                    ],
                    [[patient: 'test3'],
                    file('/Users/ebuehler/Downloads/traceback_test/test_data/P-0038647-T04-IM6_data_mutations.txt')
                    ])
                input[2] = [initial: file('/Users/ebuehler/Documents/GitHub/traceback/assets/maf_header.txt'), genotype: file('/Users/ebuehler/Documents/GitHub/traceback/assets/maf_header_genotype.txt')]
                input[3] = file('/Users/ebuehler/Downloads/traceback_test/Homo_sapiens_assembly19.fasta')
                input[4] = file('/Users/ebuehler/Downloads/traceback_test/Homo_sapiens_assembly19.fasta.fai')
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }
    }
    
    test("standard and access bams - stub") {
        options "-stub"
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [[id:'sample1', patient: 'patient1'],
                        file('sample1.bam'), 
                        file('sample1.bam.bai'), 
                        [], [], [], []
                    ],
                    [[id:'sample2', patient:'patient2'], [], [],
                        file('sample2-duplex.bam'),
                        file('sample2-duplex.bam.bai'), 
                        file('sample2-simplex.bam'),
                        file('sample2-simplex.bam.bai')
                    ])

                input[1] = Channel.of(
                    [[patient: 'patient1'], 
                    [file('patient1.txt'),]
                    ],
                    [[patient: 'patient2'], 
                    [file('patient2.maf'),]]
                    )
                input[2] = []
                input[3] = []
                input[4] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
