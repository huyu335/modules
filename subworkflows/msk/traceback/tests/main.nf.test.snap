{
    "standard and access bams - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,d8528dd2b53c5e8f814def1788328193",
                    "versions.yml:md5,f31415b0aae41057f59e929b99af930f"
                ],
                "combined_genotype_maf": [
                    
                ],
                "individual_geontype_mafs": [
                    
                ],
                "versions": [
                    "versions.yml:md5,d8528dd2b53c5e8f814def1788328193",
                    "versions.yml:md5,f31415b0aae41057f59e929b99af930f"
                ]
            }
        ],
        "timestamp": "2024-03-04T14:17:29.290228"
    },
    "sarscov2 - bam - single_end": {
        "content": [
            {
                "0": [
                    [
                        {
                            "patient": "test1"
                        },
                        [
                            "s_C_FWD2FT_X008_d-ORG-STD_genotyped.maf:md5,2be69c042bc1bf8f22ad39355dbab9de",
                            "s_C_FWD2FT_X009_d-ORG-STD_genotyped.maf:md5,2be69c042bc1bf8f22ad39355dbab9de"
                        ],
                        {
                            "patient": "test2"
                        },
                        [
                            "P-0038647-T04-IM6-ORG-STD_genotyped.maf:md5,8ff20da216b308d3b077909e3e9a92dd"
                        ],
                        {
                            "patient": "test3"
                        },
                        [
                            "C-8LCJE9-M001-d-ORG-SIMPLEX-DUPLEX_genotyped.maf:md5,8d1c582af4d677cb7740f88fa493f37e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "patient": "test1"
                        },
                        "test1_traceback.maf:md5,66eb307328c9e87fec4c0e69f075ad00"
                    ],
                    [
                        {
                            "patient": "test2"
                        },
                        "test2_traceback.maf:md5,f1ada1b73c71c4f0598bf0ace23b9713"
                    ],
                    [
                        {
                            "patient": "test3"
                        },
                        "test3_traceback.maf:md5,f2d3a63f488a352cc623e7af0cbf4929"
                    ]
                ],
                "2": [
                    "versions.yml:md5,1976e521186206f716b0eb74dd6cab1a",
                    "versions.yml:md5,d8528dd2b53c5e8f814def1788328193",
                    "versions.yml:md5,eccf6469bf87401135602b460cb4ec6c",
                    "versions.yml:md5,f31415b0aae41057f59e929b99af930f"
                ],
                "combined_genotype_maf": [
                    [
                        {
                            "patient": "test1"
                        },
                        "test1_traceback.maf:md5,66eb307328c9e87fec4c0e69f075ad00"
                    ],
                    [
                        {
                            "patient": "test2"
                        },
                        "test2_traceback.maf:md5,f1ada1b73c71c4f0598bf0ace23b9713"
                    ],
                    [
                        {
                            "patient": "test3"
                        },
                        "test3_traceback.maf:md5,f2d3a63f488a352cc623e7af0cbf4929"
                    ]
                ],
                "individual_geontype_mafs": [
                    [
                        {
                            "patient": "test1"
                        },
                        [
                            "s_C_FWD2FT_X008_d-ORG-STD_genotyped.maf:md5,2be69c042bc1bf8f22ad39355dbab9de",
                            "s_C_FWD2FT_X009_d-ORG-STD_genotyped.maf:md5,2be69c042bc1bf8f22ad39355dbab9de"
                        ],
                        {
                            "patient": "test2"
                        },
                        [
                            "P-0038647-T04-IM6-ORG-STD_genotyped.maf:md5,8ff20da216b308d3b077909e3e9a92dd"
                        ],
                        {
                            "patient": "test3"
                        },
                        [
                            "C-8LCJE9-M001-d-ORG-SIMPLEX-DUPLEX_genotyped.maf:md5,8d1c582af4d677cb7740f88fa493f37e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1976e521186206f716b0eb74dd6cab1a",
                    "versions.yml:md5,d8528dd2b53c5e8f814def1788328193",
                    "versions.yml:md5,eccf6469bf87401135602b460cb4ec6c",
                    "versions.yml:md5,f31415b0aae41057f59e929b99af930f"
                ]
            }
        ],
        "timestamp": "2024-03-04T14:54:46.814159"
    }
}